python/common/org/python/stdlib/datetime/TimeDelta.java

```diff
@@ -16,188 +40,489 @@
+                    throw new org.python.exceptions.TypeError("argument for __new__() given by name ('seconds') and position (2)");
+                }
+
+                if (kwargs.get("microseconds") != null && args.length >= 3) {
+                    throw new org.python.exceptions.TypeError("argument for __new__() given by name ('microseconds') and position (3)");
+                }
+
+                if (kwargs.get("milliseconds") != null && args.length >= 4) {
+                    throw new org.python.exceptions.TypeError("argument for __new__() given by name ('milliseconds') and position (4)");
+                }
+
+                if (kwargs.get("minutes") != null && args.length >= 5) {
+                    throw new org.python.exceptions.TypeError("argument for __new__() given by name ('minutes') and position (5)");
+                }
+
+                if (kwargs.get("hours") != null && args.length >= 6) {
+                    throw new org.python.exceptions.TypeError("argument for __new__() given by name ('hours') and position (6)");
+                }
+            }
+        }
+
+        if (args.length >= 1) {
+            addDays(getDouble("days", args[0]));
+        }
+        if (args.length >= 2) {
+            addSeconds(getDouble("seconds", args[1]));
+        }
+        if (args.length >= 3) {
+            addMicroseconds(getDouble("microseconds", args[2]));
+        }
+        if (args.length >= 4) {
+            addMilliseconds(getDouble("milliseconds", args[3]));
+        }
+        if (args.length >= 5) {
+            addMinutes(getDouble("minutes", args[4]));
+        }
+        if (args.length >= 6) {
+            addHours(getDouble("hours", args[5]));
+        }
+        if (args.length >= 7) {
+            addWeeks(getDouble("weeks", args[6]));
+        }
+
+        if (kwargs.get("days") != null) {
+            addDays(getDouble("days", kwargs.get("days")));
+        }
+        if (kwargs.get("seconds") != null) {
+            addSeconds(getDouble("seconds", kwargs.get("seconds")));
+        }
+        if (kwargs.get("microseconds") != null) {
+            addMicroseconds(getDouble("microseconds", kwargs.get("microseconds")));
+        }
+        if (kwargs.get("milliseconds") != null) {
+            addMilliseconds(getDouble("milliseconds", kwargs.get("milliseconds")));
+        }
+        if (kwargs.get("minutes") != null) {
+            addMinutes(getDouble("minutes", kwargs.get("minutes")));
+        }
+        if (kwargs.get("hours") != null) {
+            addHours(getDouble("hours", kwargs.get("hours")));
+        }
+        if (kwargs.get("weeks") != null) {
+            addWeeks(getDouble("weeks", kwargs.get("weeks")));
+        }
+
+        normalize();
+    }
+
+    private double getDouble(String key, org.python.Object object) {
+        try {
+            return ((org.python.types.Float) object.__float__()).value;
+        } catch (Exception ignored) {
```

Comment:
```suggestion
        } catch (org.python.exceptions.AttributeError ignored) {
```
It seems like that the default implementation of all these methods in `org.python.Object` throws `AttributeError`, so it should be caught directly to not accidentally catch other errors.

And the return value from `__float__` is not guaranteed to be an float, so that should be checked for as well.

These two kinds of errors will probably throw different exceptions, if only a different message.

Reply:
You are right that classes that don't implement `__float__()` will throw an `AttributeError`, however, there are also other exceptions that can happen, e.g. `ValueError` so we wanted to catch them all since there is no recovery and they should all be throwing the same exception.

But our implementation is not correct since `timedelta` does not allow e.g. `timedelta(days='1.5')` which our implementation allows so we'll change it to check that the value is an int or float instead of trying to convert it.